### PR TITLE
Gen3: Support `afterTransform` hook

### PR DESCRIPTION
**Pull request by denysoblohin-okta**
## Description:

Readme: [link](https://github.com/okta/okta-signin-widget/tree/od-afterTransform-OKTA-735653/src/v3#customizations)

Example of usage: see [hooks dir](https://github.com/okta/okta-signin-widget/tree/od-afterTransform-OKTA-735653/playground/hooks/pages)

Example of style overrides: see [customize.css](https://github.com/okta/okta-signin-widget/blob/od-afterTransform-OKTA-735653/playground/hooks/css/customize.css)

Hook examples are present in 2 versions: TS (for playground) and JS (to copy-paste into code editor in Okta admin).

Script to convert hooks example [TS](https://github.com/okta/okta-signin-widget/blob/od-afterTransform-OKTA-735653/playground/hooks/index.ts) -> [JS](https://github.com/okta/okta-signin-widget/blob/od-afterTransform-OKTA-735653/playground/hooks/index.js): `yarn --cwd playground generate:hooks:js`

PS: Question: do we still need these types? https://github.com/okta/okta-signin-widget/blob/master/src/v3/src/types/widget.ts#L136-L137

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-735653](https://oktainc.atlassian.net/browse/OKTA-735653)

### CODE REVIEW

**Review by lesterchoi-okta — COMMENTED** [review]
> PS: Question: do we still need these types? https://github.com/okta/okta-signin-widget/blob/master/src/v3/src/types/widget.ts#L136-L137

do you want to remove them or update them?

**Comment by denysoblohin-okta** [comment]
> do you want to remove them or update them?

It looks like these 2 types are not used and this confused me, so maybe comment them or remove?

File: src/v3/src/types/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+export interface TransformHookContext {
+  formName: string;
+  currentAuthenticator?: IdxAuthenticator;
+  userInfo?: UserInfo;
+  deviceEnrollment?: DeviceEnrollment;
+  nextStep?: NextStep;
+  idxContext?: IdxContext;
+}
+
+export type TransformHookFunction = (formBag: FormBag, context: TransformHookContext) => void;
```
Comment (lesterchoi-okta) [inline src/v3/src/types/hooks.ts:38]: Order of params is inconsistent with the existing event hooks

https://github.com/okta/okta-signin-widget#events

```js
signIn.on('ready', function (context) {/* ... */})
signIn.on('afterError', function (context, error) {/* ... */})
signIn.on('afterRender', function (context) {/* ... */})
```

Comment (denysoblohin-okta) [reply src/v3/src/types/hooks.ts:38]: What if we merge context and formBag in one first argument?
```js
signIn.on('afterTransform', function (context) {/* ... */})

context = {
 formBag,
 // context used for existing event hooks
 formName, authenticatorKey, methodType,
 // additional context
 userInfo, deviceEnrollment, nextStep, idxContext,
}
```

File: playground/hooks/pages/enroll-profile.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    // Customize custom boolean field (`custom_bool`) - add title, change label (with i18n support)
+    const customBoolIndex = formBag.uischema.elements.findIndex((ele): ele is FieldElement =>
+      ele.type === 'Field'
+      && (ele as FieldElement).options.type === 'boolean'
+      && (ele as FieldElement).options.inputMeta.name.includes('custom')
+    );
+    if (customBoolIndex != -1) {
+      const customBool = formBag.uischema.elements[customBoolIndex] as FieldElement;
+      customBool.translations.find(t => t.name === 'label').value = loc('custom.field.terms.label');
```
Comment (denysoblohin-okta) [inline playground/hooks/pages/enroll-profile.ts:111]: Exposed `loc` util for `afterTransform` hook and used in this sample to demonstrate i18n capabilities. 
Translations are performed before `afterTransform` hook, so setting `i18nKey` in items of `translations` array would have no effect. 
Thoughts?

**Comment by lesterchoi-okta** [comment]
@shawyu-okta mind taking a look? I've reviewed, but need to talk to docs team about the README updates

File: playground/package.json [hunk 3]
```diff
@@ -2,5 +2,8 @@
   "name": "@okta/okta-signin-widget-playground",
   "version": "1.0.0",
   "description": "## The playground App",
-  "main": "main.ts"
+  "main": "main.ts",
+  "scripts": {
+    "generate:hooks:js": "../node_modules/.bin/tsc -p ./hooks/tsconfig.json && cp -r ./hooks/js/playground/hooks/pages ./hooks && rm -rf ./hooks/js"
```
[the remark this answers is not in the document]
Comment (denysoblohin-okta) [reply playground/package.json]: Moved `tsconfig.json` to `playground` dir

**Review by trevoring-okta — APPROVED** [review]

**Pull request merged**